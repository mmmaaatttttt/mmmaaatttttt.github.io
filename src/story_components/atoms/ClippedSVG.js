import React from "react";
import PropTypes from "prop-types";

const ClippedSVG = ({ id, width, height, padding, children }) => (
  <svg xmlns="http://www.w3.org/2000/svg" viewBox={`0 0 ${width} ${height}`}>
    <defs>
      <clipPath id={`clip-path-${id}`}>
        <rect
          x={0}
          y={padding}
          height={height - 2 * padding}
          width={width - padding}
        />
      </clipPath>
    </defs>
    <g clipPath={`url(#clip-path-${id})`}>{children}</g>
  </svg>
);

ClippedSVG.propTypes = {
  id: PropTypes.string.isRequired,
  width: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
  padding: PropTypes.number.isRequired
};

ClippedSVG.defaultProps = {
  padding: 0
};

export default ClippedSVG;
