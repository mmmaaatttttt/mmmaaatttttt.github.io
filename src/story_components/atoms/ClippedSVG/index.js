import React from "react";
import styled, { css } from "styled-components";
import PropTypes from "prop-types";

const StyledSVG = styled.svg`
  display: block;
  margin-top: ${props => props.marginTop};
`;

const ClippedSVG = ({ id, width, height, marginTop, padding, children }) => {
  if (typeof padding === "number") {
    padding = {
      top: padding,
      left: 0,
      right: padding,
      bottom: padding
    };
  }
  return (
    <StyledSVG
      xmlns="http://www.w3.org/2000/svg"
      viewBox={`0 0 ${width} ${height}`}
      marginTop={marginTop}
    >
      <defs>
        <clipPath id={`clip-path-${id}`}>
          <rect
            x={padding.left}
            y={padding.top}
            height={height - padding.top - padding.bottom}
            width={width - padding.left - padding.right}
          />
        </clipPath>
      </defs>
      <g clipPath={`url(#clip-path-${id})`}>{children}</g>
    </StyledSVG>
  );
};

ClippedSVG.propTypes = {
  id: PropTypes.string.isRequired,
  width: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
  marginTop: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string
  ]).isRequired,
  padding: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.shape({
      top: PropTypes.number,
      bottom: PropTypes.number,
      left: PropTypes.number,
      right: PropTypes.number
    })
  ]).isRequired
};

ClippedSVG.defaultProps = {
  padding: 0,
  marginTop: 0,
  id: "svg"
};

export default ClippedSVG;
