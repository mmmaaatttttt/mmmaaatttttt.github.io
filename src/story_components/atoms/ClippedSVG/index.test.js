import React from "react";
import { render } from "@testing-library/react";
import ClippedSVG from "./";

it("renders without crashing", () => {
  render(<ClippedSVG />);
});

it("renders with an object for padding", () => {
  const { asFragment } = render(
    <ClippedSVG padding={{ top: 10, right: 20, bottom: 30, left: 40 }} />
  );
  expect(asFragment()).toMatchSnapshot();
});

it("matches snapshot with children", () => {
  const { asFragment } = render(
    <ClippedSVG id="svg" width={500} height={300} padding={10}>
      <rect x={0} y={10} width={100} height={200} />
      <circle cx={300} cy={100} r={100} />
    </ClippedSVG>
  );
  expect(asFragment()).toMatchSnapshot();
});
