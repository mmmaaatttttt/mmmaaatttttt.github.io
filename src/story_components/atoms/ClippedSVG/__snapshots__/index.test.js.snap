// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot with children 1`] = `
<DocumentFragment>
  <div>
    <svg
      class="sc-bdVaJa kyDpel"
      viewBox="0 0 500 300"
      xmlns="http://www.w3.org/2000/svg"
    >
      <defs>
        <clippath
          id="clip-path-svg"
        >
          <rect
            height="280"
            width="490"
            x="0"
            y="10"
          />
        </clippath>
      </defs>
      <g
        clip-path="url(#clip-path-svg)"
      >
        <rect
          height="200"
          width="100"
          x="0"
          y="10"
        />
        <circle
          cx="300"
          cy="100"
          r="100"
        />
      </g>
    </svg>
  </div>
</DocumentFragment>
`;

exports[`renders with an object for padding 1`] = `
<DocumentFragment>
  <div>
    <svg
      class="sc-bdVaJa kyDpel"
      viewBox="0 0 600 600"
      xmlns="http://www.w3.org/2000/svg"
    >
      <defs>
        <clippath
          id="clip-path-svg"
        >
          <rect
            height="560"
            width="540"
            x="40"
            y="10"
          />
        </clippath>
      </defs>
      <g
        clip-path="url(#clip-path-svg)"
      />
    </svg>
  </div>
</DocumentFragment>
`;
